src/devices/portaudioRecorder/PortAudioRecorderDeviceDriver - Copy.cpp

```diff
@@ -0,0 +1,395 @@
+            //if we return paComplete, then the callback is not called anymore.
+            //method Pa_IsStreamActive() will return 1.
+            //user needs to call Pa_StopStream() before starting a new recording session
+            finished = paComplete;
+#else
+            finished = paContinue;
+#endif
+        }
+        else
+        {
+            framesToCalc = framesPerBuffer;
+            //if we return paContinue, then the callback will be invoked again later
+            //method Pa_IsStreamActive() will return 0
+            finished = paContinue;
+        }
+
+        if( inputBuffer == nullptr )
+        {
+            for( i=0; i<framesToCalc; i++ )
+            {
+                recdata->write(0); // left
+                if(num_rec_channels == 2 ) recdata->write(0);  // right
+            }
+        }
+        else
+        {
+#if 0
+            yDebug() << "Writing" << framesToCalc*2*2 << "bytes in the circular buffer";
+#endif 
+            for( i=0; i<framesToCalc; i++ )
+            {
+                recdata->write(*rptr++);  // left
+                if(num_rec_channels == 2 ) recdata->write(*rptr++);  // right
+            }
+        }
+        return finished;
+    }
+
+    printf("No write operations requested, aborting\n");
+    return paAbort;
+}
+
+PortAudioRecorderDeviceDriver::PortAudioRecorderDeviceDriver() :
+    m_stream(nullptr),
+    m_err(paNoError),
+    m_system_resource(nullptr)
+{
+    memset(&m_inputParameters, 0, sizeof(PaStreamParameters));
+    memset(&m_dataBuffers, 0, sizeof(circularDataBuffers));
+    memset(&m_driverConfig, 0, sizeof(PortAudioRecorderDeviceDriverSettings));
+}
+
+PortAudioRecorderDeviceDriver::~PortAudioRecorderDeviceDriver()
+{
+    close();
+}
+
+
+bool PortAudioRecorderDeviceDriver::open(yarp::os::Searchable& config)
+{
+    m_driverConfig.cfg_rate = config.check("rate",Value(0),"audio sample rate (0=automatic)").asInt32();
+    m_driverConfig.cfg_samples = config.check("samples",Value(0),"number of samples per network packet (0=automatic). For chunks of 1 second of recording set samples=rate. Channels number is handled internally.").asInt32();
+    m_driverConfig.cfg_recChannels = config.check("channels", Value(0), "number of audio channels (0=automatic, max is 2)").asInt32();
+    m_driverConfig.cfg_deviceNumber = config.check("id",Value(-1),"which portaudio index to use (-1=automatic)").asInt32();
+
+    return open(m_driverConfig);
+}
+
+bool PortAudioRecorderDeviceDriver::open(PortAudioRecorderDeviceDriverSettings& config)
+{
+    m_config = config;
+
+    if (m_config.cfg_recChannels == 0)  m_config.cfg_recChannels = DEFAULT_NUM_CHANNELS;
+
+    if (m_config.cfg_rate == 0)  m_config.cfg_rate = DEFAULT_SAMPLE_RATE;
+
+    if (m_config.cfg_samples == 0) m_config.cfg_samples = m_config.cfg_rate; //  by default let's use chunks of 1 second
+
+    size_t debug_numRecBytes = m_config.cfg_samples * sizeof(SAMPLE) * m_config.cfg_recChannels;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Variable 'debug_numRecBytes' is assigned a value that is never used.](https://app.codacy.com/manual/robotology/yarp/pullRequest?prid=5162834)